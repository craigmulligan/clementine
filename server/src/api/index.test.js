const { app } = require('../index')
const proto = require('apollo-engine-reporting-protobuf')
const zlib = require('zlib')
const promisify = require('util').promisify
const gzip = promisify(zlib.gzip)
const { Trace, Graph, User, sql, db } = require('../persistence')

beforeEach(() => {
  return db.query(sql`START TRANSACTION`)
})
afterEach(() => {
  return db.query(sql`ROLLBACK`)
})

function formatProto(path) {
  const messageJSON = require(path)
  const message = proto.FullTracesReport.fromObject(messageJSON)
  const buffer = proto.FullTracesReport.encode(message).finish()
  return gzip(buffer)
}

describe('/api/ingress', () => {
  test('No ApiKey', async () => {
    const compressed = await formatProto('./__data__/traces.json')
    const request = require('supertest').agent(app)

    // TODO create graph
    await request
      .post('/api/ingress/traces')
      .set('content-encoding', 'gzip')
      .send(compressed)
      .expect(403)
  })

  test('uncompressed', () => {
    // should handle uncompressed payloads too
  })

  test('Happy path', async () => {
    const compressed = await formatProto('./__data__/traces.json')
    const request = require('supertest').agent(app)

    const user = await User.create('email@email.com', '123')
    const graph = await Graph.create('myGraph', user.id)

    // TODO create graph
    await request
      .post('/api/ingress/traces')
      .set('content-encoding', 'gzip')
      .set('x-api-key', graph.id + ':123')
      .send(compressed)
      .expect(201)

    const traces = await Trace.findAll({ graphId: graph.id })
    expect(traces.length).toBe(2)
    const t = traces[0]
    expect(t).toHaveProperty('id')
    expect(t).toHaveProperty('graphId')
    expect(t).toHaveProperty('duration')
    expect(t).toHaveProperty('startTime')
    expect(t).toHaveProperty('endTime')
    expect(t).toHaveProperty('root')
    expect(t).toHaveProperty('clientName')
    expect(t).toHaveProperty('clientVersion')
  })

  test('happy path - with errors', async () => {
    const compressed = await formatProto('./__data__/traces.json')
    const request = require('supertest').agent(app)

    const user = await User.create('email@email.com', '123')
    const graph = await Graph.create('myGraph', user.id)

    // TODO create graph
    await request
      .post('/api/ingress/traces')
      .set('content-encoding', 'gzip')
      .set('x-api-key', graph.id + ':123')
      .send(compressed)
      .expect(201)

    const traces = await Trace.findAll({ graphId: graph.id })
    expect(traces.length).toBe(2)
    const t = traces[0]
    expect(t).toHaveProperty('id')
    expect(t).toHaveProperty('graphId', graph.id)
    expect(t).toHaveProperty('duration')
    expect(t).toHaveProperty('startTime')
    expect(t).toHaveProperty('endTime')
    expect(t).toHaveProperty('root')
    expect(t).toHaveProperty('clientName')
    expect(t).toHaveProperty('clientVersion')
  })
})
