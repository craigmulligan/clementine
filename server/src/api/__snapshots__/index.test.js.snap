// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/ingress Happy path 1`] = `
Array [
  Object {
    "clientname": null,
    "clientversion": null,
    "duration": 67729237,
    "endtime": 2020-03-27T12:34:26.000Z,
    "graphid": "1b584ccc-aa2a-4fb8-997f-90b030f8e6a3",
    "id": "448910ef-9511-4fab-85bd-ddf6a0c71125",
    "key": "# GET_USER
query GET_USER{user{email id}}",
    "root": Object {
      "cachePolicy": null,
      "child": Array [
        Object {
          "cachePolicy": null,
          "child": Array [],
          "endTime": "65005807",
          "error": Array [],
          "id": "responseName",
          "originalFieldName": "",
          "parentType": "Query",
          "responseName": "user",
          "startTime": "29032827",
          "type": "User",
        },
      ],
      "endTime": "0",
      "error": Array [],
      "originalFieldName": "",
      "parentType": "",
      "startTime": "0",
      "type": "",
    },
    "starttime": 2020-03-27T12:35:35.000Z,
  },
  Object {
    "clientname": null,
    "clientversion": null,
    "duration": 4679795,
    "endtime": 2020-03-27T12:32:03.000Z,
    "graphid": "1b584ccc-aa2a-4fb8-997f-90b030f8e6a3",
    "id": "79d7df7d-3c1c-46b3-919e-bf0105a65bf7",
    "key": "# GET_USER
query GET_USER{user{email id}}",
    "root": Object {
      "cachePolicy": null,
      "child": Array [
        Object {
          "cachePolicy": null,
          "child": Array [],
          "endTime": "4302402",
          "error": Array [],
          "id": "responseName",
          "originalFieldName": "",
          "parentType": "Query",
          "responseName": "user",
          "startTime": "1037483",
          "type": "User",
        },
      ],
      "endTime": "0",
      "error": Array [],
      "originalFieldName": "",
      "parentType": "",
      "startTime": "0",
      "type": "",
    },
    "starttime": 2020-03-27T12:32:08.000Z,
  },
]
`;
